src/script/script.h

```diff
@@ -381,6 +382,112 @@ class CScriptNum
     int64_t m_value;
 };
 
+// A helper class for creating, reading, and writing the multisig hint
+// field placed in the extra stack position of a CHECKMULTISIG when
+// SCRIPT_VERIFY_REQUIRE_VALID_SIGNATURES is in effect.
+class MultiSigHint
+{
+protected:
+    // A bitset is used to indicate which keys do NOT have signatures
+    // present, and should be skipped over. Keys for which there is a
+    // signature present have their corresponding bit clear.
+    //
+    // The ordering of bits within the hint is the same as the
+    // ordering of the pubkeys on the stack: if there are N public
+    // keys, the 1st public key (closest to the top of the stack) is
+    // indicated by the lowest-order bit in the hint; use of the Nth
+    // public key (nearest the bottom of the stack) is indicated with
+    // the Nth bit. For example:
+    //
+    //     <hint> <sigB> <sigA> 2 <pubC> <pubB> <pubA> 3 CHECKMULTISIG
+    //     hint:               0b     1      0      0  = 4
+    //
+    // Since the hint is pushed onto the stack as an integer, it can
+    // be more space efficient to drop the signature(s) corresponding
+    // to the last keys of a threshold when all keys are available.
+    // For example, an 11-of-15 threshold has a hint value of 0x000f
+    // when it is the first 11 public keys that are used. This is
+    // compactly pushed onto the stsack with a single byte, the OP_16
+    // opcode.
+    std::bitset<MAX_PUBKEYS_PER_MULTISIG> skipped_keys;
+    // The skipped_keys bitset is setup to store bits for 20 keys, the
+    // maximum number of pubkeys allowed in a single CHECKMULTISIG.
+    // We need the actual number of keys in order to calculate how
+    // many signatures should be present, since higher order bits of
+    // the bitfield are not significant.
+    int num_keys;
+
+public:
+    // By default no bits are set, which means ALL keys are used. This
+    // can only be valid for a n-of-n multisig. However a reasonable
+    // way to use this class is to initialize a hint object with this
+    // default value and then call skip_key() for the pubkeys which
+    // are not used.
+    MultiSigHint(int _num_keys)
+        : skipped_keys(0)
+        , num_keys(_num_keys)
```

Comment:
Recent code style is to give member vars prefix `m_`. That lets you give constructor params the non-prefixed name as is, which looks a little cleaner.